src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs

```diff
@@ -374,7 +388,34 @@ private int RunCore(TextWriter consoleOutput, ErrorLogger errorLogger, Cancellat
                     var analyzerOptions = new AnalyzerOptions(ImmutableArray<AdditionalText>.CastUp(additionalTextFiles));
 
                     analyzerDriver = AnalyzerDriver.CreateAndAttachToCompilation(compilation, analyzers, analyzerOptions, analyzerManager, addExceptionDiagnostic, Arguments.ReportAnalyzer, out compilation, analyzerCts.Token);
-                    getAnalyzerDiagnostics = () => analyzerDriver.GetDiagnosticsAsync(compilation).Result;
+                    reportAnalyzer = Arguments.ReportAnalyzer && !analyzers.IsEmpty;
+                }
+
+                Compilation newCompilation = null;
+                if ((analyzerDriver != null) || includeGenerators)
+                {
+                    getAnalyzerDiagnostics = (diags, result) =>
+                    {
+                        if (result && (analyzerDriver != null))
+                        {
+                            var hostDiagnostics = analyzerDriver.GetDiagnosticsAsync(compilation).Result;
+                            diags.AddRange(hostDiagnostics);
+                            if (hostDiagnostics.Any(x => x.Severity == DiagnosticSeverity.Error))
+                            {
+                                result = false;
+                            }
+                        }
+                        if (includeGenerators)
+                        {
+                            newCompilation = AddGeneratedSource(compilation, sourceGenerators, cancellationToken);
```

Comment:
Looks fishy that you are assigning this outer variable inside this lamba function.